src/cuda/cuda_sampling.cu

```diff
@@ -548,15 +548,26 @@ void GetTopKSubset(SamplingData* data, cudaStream_t stream, float* scores_in, fl
                              k,                          \
                              temperature);
 
-  if (k <= 4) {
+  bool use_sort_and_pick_topk = false;
+
+  // TODO(hasesh): Tweak this heuristic by performing more benchmarks.
+  // It is likely that the sort and pick top_k approach works well for
+  // smallish batch_sizes on most hardware. For now, we limit it to
+  // sampling (batch_size == 1) && when vocab_size is "pretty large"
+  // (also needs benchmarking to identify cut-off) && when top_k is >= 20.
+  if (batch_size == 1 && vocab_size >= 100000 && k >= 20) {
+    use_sort_and_pick_topk = true;
+  }
+
+  if (!use_sort_and_pick_topk && k <= 4) {
     GetTopK(4);
-  } else if (k <= 8) {
+  } else if (!use_sort_and_pick_topk && k <= 8) {
     GetTopK(8);
-  } else if (k <= 16) {
+  } else if (!use_sort_and_pick_topk && k <= 16) {
     GetTopK(16);
-  } else if (k <= 32) {
+  } else if (!use_sort_and_pick_topk && k <= 32) {
     GetTopK(32);
-  } else if (k <= 64) {
+  } else if (!use_sort_and_pick_topk && k <= 64) {
```

Comment:
For `GetTopK`, there is a `scores_out[i + batch * k] = top_k_sequence.value[i] / temperature;` in 
 https://github.com/microsoft/onnxruntime-genai/blob/b6d74e2e7efe2ae4d2b4eda7a10bcbd2eb3b1ff3/src/cuda/cuda_sampling.cu#L469, seems we need change in `LaunchSort`